### ISSUE TITLE
Schedules repeat multiple times BEFORE the scheduled time

### ISSUE THREAD

**Issue by yashendra2797** [pterodactyl/panel#3351]
**STOP: READ FIRST, AND THEN DELETE THE ABOVE LINES**

**Background (please complete the following information):**

* Panel or Wings: Panel
* Version of Panel/Wings: 1.4.1 (Panel) | 1.4.2 (Wings)
* Panel Logs: Last log is from 06 May (Screenshot: https://yash.gg/3fePeCx). Log in question: http://bin.ptdl.co/zw85x
* Wings Logs: https://ptero.co/wigefavehu
* Server's OS: Ubuntu 20.04.2 LTS (GNU/Linux 5.4.0-72-generic x86_64) (Panel and Wings are on separate machines, but same OS)
* Your Computer's OS & Browser: macOS Big Sur v11.2.3 (20D91) | Safari Version 14.0.3 (16610.4.3.1.7)

**Describe the bug**
I have 2 separate schedules. one for Restarting panel, one for Backing it up to B2. They are an hour apart. Restart cron is `0 4 * * *`, while Backup cron is `0 5 * * *`. I checked their validity from [Crontab.guru](https://crontab.guru/#0_4_*_*_*) They both have just the relevant item inside, nothing else. Screenshot: https://yash.gg/2PYcn3L

Servers where I have these schedule, start restarting at 03:30 instead of 4, and they will **keep restarting in a loop** until 4. Then they'll be fine till the next day. Similarly, my backups start happening at 04:30 instead of 5, and will keep happening, until 5. 

Here's a screenshot of one of my servers backing up 6 times yesterday: https://yash.gg/3oaEzNg

Here's a screenshot of 31 log.gz files being generated as my server restarted 31 times: https://yash.gg/33Ctw6d

Now if it was just a restart issue, I would've said, hey maybe my java is messed up or server is crashing. But, this happens on backups as well, which are only controlled by panel. This is why I made a separate schedule for Backups an hour apart to test.

If it was just the schedules running 30 minutes earlier, I would've said its a time zone issue.

But my schedules run 30 minutes early, and will run on a loop until the last schedule fires off on the scheduled time.

This happens on every server I have tried this schedule in.

This is needless to say, extremely weird.

Here's pteroq logs found via `journalctl --unit pteroq --since 2021-05-11`. I copied logs from 5 am on the 11th when the schedule stops to the latest line. You can see how pteroq executes multiple schedules non stop from exactly 330 am. to exactly 5am. I don't have a single schedule for 330 AM: https://paste.lucko.me/yHeC4htrAs

I have spent two months troubleshooting this. Discord was no help until today when I was told to submit an issue.

I have checked system date using `date` on both panel and wings, checked php date, and mysql date on my panel. They all sync up.

I am desperate, and willing to provide any logs or do anything to help with this. I am at a loss and willing to whatever it takes.

Please provide additional information too, depending on what you have issues with:
Panel: `php -v` (the php version in use): PHP 8.0.5 (cli) (built: May  3 2021 11:30:57) ( NTS )
Wings: `uname -a` and `docker info` (your kernel version and information regarding docker)
`Linux mediocre.gg 5.4.0-72-generic #80-Ubuntu SMP Mon Apr 12 17:35:00 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`
Panel docker info: https://paste.lucko.me/obJj9ECvBS
Wings docker info: https://paste.lucko.me/Ep7LheqzAZ

**To Reproduce**
Steps to reproduce this behavior:

1. Create a schedule for restart with `0 4 * * *` as cron
2. Create another schedule for backup to B2 wwith `0 5 * * *` as cron
3. Wait until the time for schedule rolls around
4. Watch from 03:30 as your servers start restarting in a loop

**Expected behavior**
Schedule for each server fires on time, and only once.

**Comment by DaneEveritt** [comment]
Honestly, couldn't even begin to take a stab at what would be causing this. It definitely is not affecting other people to my knowledge, so I'm almost certain you have some weird setup going on. Do you have multiple instances of the Panel running, or multiple schedule processors? There is no reason a scheduled task would keep repeating like that, it should just fail gracefully and re-run at the next interval, or succeed, and run at the next interval.

Also, the fact that they're running 30 minutes early really points (initially without any other details) to multiple instances running, and all of those instances having different system and database times.

**Comment by yashendra2797** [comment]
> Honestly, couldn't even begin to take a stab at what would be causing this. It definitely is not affecting other people to my knowledge, so I'm almost certain you have some weird setup going on. Do you have multiple instances of the Panel running, or multiple schedule processors? There is no reason a scheduled task would keep repeating like that, it should just fail gracefully and re-run at the next interval, or succeed, and run at the next interval.
> 
> Also, the fact that they're running 30 minutes early really points (initially without any other details) to multiple instances running, and all of those instances having different system and database times.

To my knowledge I am not running multiple panels, but could you please let me know how to check if I have multiple instances of panel and pteroq running?

**Comment by yashendra2797** [comment]
@DaneEveritt Hey sorry to tag you, but its been 23 days can you please let me know how to troubleshoot this in reference to my comment above?

**Comment by yashendra2797** [comment]
I used `pgrep` to check, I only found one instance of wings running. I also run `sudo systemctl list-units --type service`, and only found one `pteroq.service` in the list. Here's my crontab:

```
# Edit this file to introduce tasks to be run by cron.
# 
# Each task to run has to be defined through a single line
# indicating with different fields when the task will be run
# and what command to run for the task
# 
# To define the time you can provide concrete values for
# minute (m), hour (h), day of month (dom), month (mon),
# and day of week (dow) or use '*' in these fields (for 'any').
# 
# Notice that tasks will be started based on the cron's system
# daemon's notion of time and timezones.
# 
# Output of the crontab jobs (including errors) is sent through
# email to the user the crontab file belongs to (unless redirected).
# 
# For example, you can run a backup of all your user accounts
# at 5 a.m every week with:
# 0 5 * * 1 tar -zcf /var/backups/home.tgz /home/
# 
# For more information see the manual pages of crontab(5) and cron(8)
# 
# m h  dom mon dow   command

* * * * * php /var/www/pterodactyl/artisan schedule:run >> /dev/null 2>&1
52 0 * * * "/etc/letsencrypt"/acme.sh --cron --home "/etc/letsencrypt" > /dev/null
* * * * * bash /etc/hetrixtools/hetrixtools_agent.sh >> /etc/hetrixtools/hetrixtools_cron.log 2>&1
```

Here's my `pteroq.service` file:

```
# Pterodactyl Queue Worker File
# ----------------------------------

[Unit]
Description=Pterodactyl Queue Worker
After=redis-server.service

[Service]
# On some systems the user and group might be different.
# Some systems use `apache` or `nginx` as the user and group.
User=www-data
Group=www-data
Restart=always
ExecStart=/usr/bin/php /var/www/pterodactyl/artisan queue:work --queue=high,standard,low --sleep=3 --tries=3

[Install]
WantedBy=multi-user.target
```

I have also attached a copy of the content of my` /etc/systemd/system` file:

![here](https://dsc.cloud/yashendra2797/worksheets-enfoulder-nitridization-nonelastically/Screen-Shot-2021-06-15-05-21-56.89.jpg)


Please let me know what else I can do to help.

**Comment by sscs12345** [comment]
Need help on the same exact issue attaching my `ptero diagnostics`
[https://ptero.co/axepixenoh](https://ptero.co/axepixenoh)

Please fix it 👍

**Comment by trenutoo** [comment]
> Need help on the same exact issue attaching my `ptero diagnostics`
> https://ptero.co/axepixenoh
> 
> Please fix it 👍

Wings have no part in the schedules. That'd be the Panel machine and Pteroq queue worker.

This is possibly a duplicate of #2075

**Comment by adsbhatial** [comment]
I can confirm this is happening in pterodactyl v1.5.1. I have checked the logs but can't find anything useful at all.  I am running a single server with a single node. I have checked timezones of my panel, wings, OS and php. All set to Asia/Kolkata. 

My system info 

uname -a                                                                         :
`Linux oci-vm1 5.11.0-1016-oracle #17~20.04.1-Ubuntu SMP Thu Aug 12 06:20:07 UTC 2021 aarch64 aarch64 aarch64 GNU/Linux`
System Modules                                                             : [https://pastebin.com/AHP3NfdH](https://pastebin.com/AHP3NfdH)
Panel timezone(`/var/www/pterodactyl/.env`)                :  `APP_TIMEZONE=Asia/Kolkata` 
Wings timezone(`/etc/pterodactyl/config.yml`)              : `timezone: Asia/Kolkata`
PHP timezone('etc/php/7.4/apache2/php.ini')                : `Default timezone | Asia/Kolkata`
Mysql Timezone(`SELECT @@system_time_zone as tz;`) : `IST` 
Schedules                                                                         : [https://i.imgur.com/3lGNSIf.png](https://i.imgur.com/3lGNSIf.png)

journalctl for pteroq.service                                             : 
```
Aug 27 11:30:04 oci-vm1 php[169497]: [2021-08-27 11:30:04][314] Processing: Pterodactyl\Jobs\Schedule\RunTaskJob
Aug 27 11:30:04 oci-vm1 php[169497]: [2021-08-27 11:30:04][314] Processed:  Pterodactyl\Jobs\Schedule\RunTaskJob
Aug 27 11:31:01 oci-vm1 php[169497]: [2021-08-27 11:31:01][315] Processing: Pterodactyl\Jobs\Schedule\RunTaskJob
Aug 27 11:31:01 oci-vm1 php[169497]: [2021-08-27 11:31:01][315] Processed:  Pterodactyl\Jobs\Schedule\RunTaskJob
Aug 27 11:32:01 oci-vm1 php[169497]: [2021-08-27 11:32:01][316] Processing: Pterodactyl\Jobs\Schedule\RunTaskJob
Aug 27 11:32:01 oci-vm1 php[169497]: [2021-08-27 11:32:01][316] Processed:  Pterodactyl\Jobs\Schedule\RunTaskJob
Aug 27 11:33:05 oci-vm1 php[169497]: [2021-08-27 11:33:05][317] Processing: Pterodactyl\Jobs\Schedule\RunTaskJob
Aug 27 11:33:05 oci-vm1 php[169497]: [2021-08-27 11:33:05][317] Processed:  Pterodactyl\Jobs\Schedule\RunTaskJob
Aug 27 11:34:02 oci-vm1 php[169497]: [2021-08-27 11:34:02][318] Processing: Pterodactyl\Jobs\Schedule\RunTaskJob
Aug 27 11:34:02 oci-vm1 php[169497]: [2021-08-27 11:34:02][318] Processed:  Pterodactyl\Jobs\Schedule\RunTaskJob
```
ptreroq.service file

```
# Pterodactyl Queue Worker File
# ----------------------------------

[Unit]
Description=Pterodactyl Queue Worker
After=redis-server.service

[Service]
# On some systems the user and group might be different.
# Some systems use `apache` or `nginx` as the user and group.
User=www-data
Group=www-data
Restart=always
ExecStart=/usr/bin/php /var/www/pterodactyl/artisan queue:work --queue=high,standard,low --sleep=3 --tries=3
StartLimitInterval=180
StartLimitBurst=30
RestartSec=5s

[Install]
WantedBy=multi-user.target
```
There are no logs in the `/var/www/pterodactyl/storage/logs/ ` for current date

**Comment by yashendra2797** [comment]
> I have checked timezones of my panel, wings, OS and php. All set to Asia/Kolkata.

Wait I'm in Asia/Kolkata as well. Interesting.

**Comment by DaneEveritt** [comment]
I'm closing this — I don't have the bandwidth to dig into such a highly specific issue impacting a tiny portion of users.

**Comment by Icingsgg** [comment]
Hi Dane, this issue still exists, even though the Time zone for Panel, PHP & MySQL are same. Can you help us here with a workaround?

**Comment by yashendra2797** [comment]
@adsbhatial @Icings2003 Do you both by any chance have the time zone set to Asia/Kolkata? You both seem Indian just like me. Maybe this issue is isolated to IST Time Zone and that's why its not widespread?

**Comment by Icingsgg** [comment]
> @adsbhatial @Icings2003 Do you both by any chance have the time zone set to Asia/Kolkata? You both seem Indian just like me. Maybe this issue is isolated to IST Time Zone and that's why its not widespread?

Yes I'm also using IST.

**Comment by yashendra2797** [comment]
> 

I have a suspicion that this is an IST specific issue as in the last 1 year the only people who have commented saying they're facing this issue are Indians. I will try switching to GMT tonight if I have the time and report back.

**Comment by yashendra2797** [comment]
Update:

I changed the timezone on my Panel server, and 2 wings servers. I left timezone as Asia/Kolkata on my 3rd wing server. On server 1, restart and backup was successful. On 2nd server (which was the wings server for which I raised this issue) after time zone change to UTC, the schedules were successful. On my 3rd relatively new server, which had timezone of Asia/Kolkata, restart was fine, backup failed.

On top of this, on my panel, the schedules run on time, but last run shows 30 mins ago of my local time. So, env and server is set to UTC, my timezone is +05:30, but last run is shown as +05:00,

And in journalctl, the time for the pteroq event is in IST even though server and display is UTC.

Truly bizarre bug that I cannot figure it out.
